sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableConfig.java

```diff
@@ -0,0 +1,231 @@
+
+import com.google.auto.value.AutoValue;
+import com.google.cloud.bigtable.config.BigtableOptions;
+import com.google.cloud.bigtable.config.CredentialOptions;
+import org.apache.beam.sdk.extensions.gcp.options.GcpOptions;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.options.ValueProvider;
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.apache.beam.sdk.transforms.display.DisplayData;
+
+import javax.annotation.Nullable;
+import java.io.Serializable;
+
+/**
+ * Configuration for a Cloud Bigtable client.
+ */
+@AutoValue
+abstract class BigtableConfig implements Serializable {
+
+  /**
+   * Modification function for Bigtable Options.
+   */
+  abstract SerializableFunction<BigtableOptions, BigtableOptions> getBigtableOptionsModificationFunction();
+
+  /**
+   * Returns the options for Bigtable.
+   */
+  @Nullable
+  abstract ValueProvider<BigtableOptions> getBigtableOptions();
+
+  /**
+   * Returns the project id being written to.
+   */
+  @Nullable
+  abstract ValueProvider<String> getProjectId();
+
+  /**
+   * Returns the innstance id being written to.
+   */
+  @Nullable
+  abstract ValueProvider<String> getInstanceId();
+
+  /**
+   * Returns the table id being written to.
+   */
+  @Nullable
+  abstract ValueProvider<String> getTableId();
+
+  /**
+   * Weather validate that table exists before writing.
+   */
+  @Nullable
+  abstract ValueProvider<Boolean> getValidation();
+
+  /**
+   * For testing purposes only.
+   */
+  @Nullable
+  abstract BigtableService getBigtableService();
+
+  abstract Builder toBuilder();
+
+  static BigtableConfig create() {
+    return builder().build();
+  }
+
+  static Builder builder() {
+    return new AutoValue_BigtableConfig.Builder();
+  }
+
+  void validate() {
+    checkNotNull(
+        getTableId(),
+        "BigtableIO requires table id to be set with withTableId method");
+    checkArgument(getTableId().isAccessible(), "bigtable table is not accessible");
+    checkNotNull(getTableId().get(), "bigtable options could not be null");
+
+    boolean bigtableOptionsAvailable = getBigtableOptions() != null && getBigtableOptions().isAccessible();
+    boolean projectIdAvailable = getProjectId() != null && getProjectId().isAccessible();
+    boolean instanceIdAvailable = getInstanceId() != null && getInstanceId().isAccessible();
+
+    checkArgument(!projectIdAvailable
+            && (!bigtableOptionsAvailable || getBigtableOptions().get().getProjectId() == null),
+        "Could not obtain project id");
+
+    checkArgument(!instanceIdAvailable
+            && (!bigtableOptionsAvailable || getBigtableOptions().get().getInstanceId() == null),
+        "Could not obtain instance id");
+  }
+
+  void populateDisplayData(DisplayData.Builder builder) {
+    builder
+        .addIfNotNull(DisplayData.item("projectId", getProjectId()).withLabel("Project ID"))
```

Comment:
I think that getProjectIdHelper() might make sense here... it could check either getProjectId() or getBigtableOptions().getProjectId() to get the right value... same with instanceId()